Fix incorrect data interval alignment due to assumption on input time alignment

wanlce opened the pull request:
This introduces a new `_align_to_prev` util function for cron-based interval alignment. Previously we assumed `_get_prev(_align_to_next(t))` is enough for this purpose, but this proved incorrect (see discussions in #21715), and an additional alignment logic is needed to avoid the inferred interval being off by one period when the input value has unexpected alignment (unexpectedly aligning for a manual run, and unexpected _not aligning_ for an auto one).

See also #21715 
Fix #23689

uranusjr commented on airflow/timetables/interval.py:
```diff
@@ -58,6 +58,10 @@ def _align(self, current: DateTime) -> DateTime:
         raise NotImplementedError()
 
+    def _align_to_prev(self, current: DateTime) -> DateTime:
+        """Align given time to the scheduled."""
```
Please

1. Add more description on what this function does (similar to the docstring for `_align_to_next`)
2. Describe _why_ this is needed and how it’s different from `self._prev(self._align_to_next(t))` (when a DAG’s schedule changes, as you mentioned in the original PR).

uranusjr commented on airflow/timetables/interval.py:
```diff
@@ -76,21 +80,23 @@ def next_dagrun_info(
                 return None
             start = earliest
-        else:  # There's a previous run.
+        else:  # There's a previous run
```
Accidental?

wanlce replied:
sorry !!!   Thank you for your review, I made changes. Looking forward to your new comment.

wanlce replied:
Hello

Commit:
Fix incorrect data_interval_start due to DAG's schedule changed

This PR fixes the incorrect first run of data_interval_start after
changing the scheduling time.

* Added _align_to_prev function for scheduling alignment after time change.

* renamed _align to _align_to_next.

uranusjr commented on airflow/timetables/interval.py:
```diff
@@ -58,6 +58,21 @@ def _align(self, current: DateTime) -> DateTime:
+        changed, _algin_to_next does not correct the scheduling time in time,
+        resulting in the first scheduling after the scheduling time has been
+        changed remaining the same. So we would need to align forward to ensure
+        that it works correctly after a scheduling time change.
```
```suggestion
        It is not enough to use ``_get_prev(_align_to_next())`` instead. When a
        DAG's schedule changes, this alternative would make the first scheduling
        after the schedule change remain the same. We need to align forward to
        ensure it works correctly in this situation.
```

uranusjr committed:
Wrap function name with backticks

uranusjr committed:
Merge branch 'main' into fix-timetables-interval-start

uranusjr committed:
Add test case for auto run alignment

uranusjr committed:
Fix manual run alignment
